components/raftstore/src/store/fsm/peer.rs

```diff
@@ -4683,17 +4701,72 @@ where
                 "epoch" => ?region_epoch,
                 "current_epoch" => ?region.get_region_epoch(),
             );
+            // test purpose
+            test_only_callback(cb, self.fsm.peer.region_buckets.clone());
             return;
         }
-        let mut region_buckets = Buckets::default();
-        region_buckets.version = timespec_to_ns(monotonic_raw_now());
-        region_buckets.set_keys(bucket_keys.into());
-        // add region's start/end key
-        region_buckets
-            .keys
-            .insert(0, region.get_start_key().to_vec());
-        region_buckets.keys.push(region.get_end_key().to_vec());
-        self.fsm.peer.region_buckets = region_buckets;
+
+        if let Some(bucket_ranges) = bucket_ranges {
+            assert_eq!(buckets.len(), bucket_ranges.len());
+            let mut region_buckets_keys = self.fsm.peer.region_buckets.keys.to_vec();
+            let mut region_buckets = Buckets::default();
+            region_buckets.version = timespec_to_ns(monotonic_raw_now());
+            let mut i = 0;
+            let mut j = 0;
+            let buckets = buckets.drain(..);
+            for mut bucket in buckets {
+                while i < region_buckets_keys.len() {
+                    if region_buckets_keys[i] == bucket_ranges[j].0 {
+                        // the bucket size is small and does not have split keys,
+                        // then it should be merged with its left neighbor
+                        if bucket.keys.is_empty()
+                            && bucket.size
+                                <= self.ctx.coprocessor_host.cfg.region_bucket_merge_size.0
```

Comment:
So 128 buckets `[128MiB, 1MiB, ..., 1MiB]` will be merged as `[255MiB]` instead of `[128MiB, 128MiB]`?

Reply:
but as long as 255MiB increases to 256MiB, then it will split into [128MiB, 128MiB].  Here we have the cap of unbalance.

Reply:
When will it be split? I don't see further check is done after the loop.

Reply:
> When will it be split? I don't see further check is done after the loop.

The next time the check split run.

Reply:
Refined a bit the logic. Now it will only do the merge if the combined bucket size does not exceed 2x of region bucket size. This is to avoid too big bucket size.
In your example, it's still 255MiB.